### ISSUE TITLE
Should there be an EVP_PKEY_todata()?

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#14657]
We have the API `EVP_PKEY_fromdata()` which creates a new key based on a set of OSSL_PARAMs. It is possible to get param data from an existing key using `EVP_PKEY_get_params`. However this requires you to know up front all the params data that you might want to get and to set up the OSSL_PARAM array accordingly. It would be nice to have an `EVP_PKEY_todata()` function which creates an OSSL_PARAM array for you and populates it with all the relevant parameters required for a subsequent `EVP_PKEY_fromdata()` call to be successful.

**Comment by paulidale** [comment]
Why not use the output from the `EVP_PKEY_gettable_params` call to get the params?
The gettable params would need to be cloned to avoid them being read only.

**Comment by mattcaswell** [comment]
Are all required parameters (sufficient to re-create the key using `EVP_PKEY_fromdata()`) exposed by `EVP_PKEY_gettable_params`?

**Comment by t8m** [comment]
> Are all required parameters (sufficient to re-create the key using `EVP_PKEY_fromdata()`) exposed by `EVP_PKEY_gettable_params`?

I'd consider that to be a bug in the provider if they aren't.

There is however a problem with usability of the simple "cloning" - for bignum parameters you do not know how big the data will be in advance and you'll have to call the getparams twice with preallocations. This is clearly visible in the EVP_PKEY_get_bn_param() implementation.

**Comment by levitte** [comment]
The internal export function passes a callback, which is passed an OSSL_PARAM array constructed by the provider. Just wrap that with a public API, where the callback comes from the caller. Done deal.

**Comment by paulidale** [comment]
I think it would be more user/application friendly, to clone the OSSL_PARAM array for the user -- gets the memory allocated on the proper side of the provider boundary, lets the app take control of the life time etc.

**Comment by levitte** [comment]
I think that duplicating the OSSL_PARAM array should be a decision for the caller, not a unilateral decision that we make for them.

**Comment by slontis** [comment]
You need to look at what the export does before making this comment.
By the time the export has finished the params returned from the  callback are destroyed.

**Comment by paulidale** [comment]
I agree with @slontis here.  We need persistence,

**Comment by levitte** [comment]
> You need to look at what the export does before making this comment.

I'm fully aware how it works.

> By the time the export has finished the params returned from the callback are destroyed.

The params *recieved by* the callback only exist until the callback ends, yes.  And?
Could there be a misunderstanding here?  When I said that we could simply make the currently internal export function (`evp_keymgmt_util_export()`) public, I really meant that, and that means that the application would provide the callback.  Therefore, it would be entirely up to the application what they want to do with the params that the callback receives.

**Comment by levitte** [comment]
In other words, I was seeing this:

``` C
int EVP_PKEY_todata(const EVP_PKEY *pk, int selection,
                    OSSL_CALLBACK *export_cb, void *export_cbarg);
```

But, from seeing discussions here and in #14785, it looks like we have very different ideas of what `EVP_PKEY_todata()` should offer.  Would you mind showing your idea in a bit more concrete form?

**Comment by slontis** [comment]
> I'm fully aware how it works.

What you have above is fine.. I want the param data to be available after EVP_PKEY_todata() has been called. i.e within the 
export_cb the user can use the OSSL_PARAM_dup() if they want to.

**Comment by t-j-h** [comment]
@levitte that is a pretty poor interface compared to the fromdata (other side).
That interface existing doesn't hurt to have (and exposed) - but we should be able to offer a simple interface that returns a usable OSSL_PARAM array and expect our users to actually use that interface in preference to having to run an iterator and arrange collection themselves.

**Comment by levitte** [comment]
I'm fine with both.  We could expose `evp_pkey_util_export()` as `EVP_PKEY_export()` if that makes everyone feel better.

**Comment by paulidale** [comment]
A callback is a dreadful API for this.  An OSSL_PARAM array is what we accept as input, we should be able to output one.  It is much friendlier and easier to use.

**Comment by levitte** [comment]
I would still like to offer the *choice* to the user / application.  We do not know what the resulting OSSL_PARAM will be used for, but you can be sure that there are some who'd prefer not to waste more memory allocation and copying than necessary.

Can we settle on both?

**Comment by paulidale** [comment]
Both is okay since we've got to live with the callback form internally, it shouldn't be a major maintenance burden.

**Comment by t8m** [comment]
Closed as #14800 is merged.
